File: Biot.jl-main/src/MakeSolenoid.jl


"""
This function makes an elliptic solenoid and calculates the inductance.
    The function calculates the flux through the center of each loop's cross-section
        Lₛ = Φ/I (self inductance = flux per amp)
        Lₘ₂₁ = (Φ₂₁ ⋅ Φ₂₂/|Φ₂₂|) /I
            -mutual inductance from the 1ˢᵗ wire on 2ⁿᵈ is flux from 1 in the 2ⁿᵈ's cross section
            that is aligned with the flux from the second on itself. All per ampere
        Lₘ₂₁ = Lₘ₁₂
        Total L = ∑ᵢ∑ⱼ(Lᵢⱼ)
    The inputs are:
    N - Number of turns
    r₁ - one radius of the ellipse
    r₂ - Second radius of ellipse
    L - length of solenoid (axially)
    kwargs:
    MeasLayers - Number of concentric layers to calc field
                    The field is calculated over concentric ellipses (linearly spaced)
                    This is the number of concentric test point layers
    NLayers -  Number of test points per layer
    NPtsPath  - number of points that each wire loop will be discretized into,
    PlotOn deternimes if the coil will be plotted as it builds

"""
function eval_Solenoid_Induct(
    N, #Number of turns
    r₁, #Radius 1 of ellipse
    r₂,#Radius 2 of ellipse
    L; #Length of solenoid in meters
    DownSampleFac=1,PlotOn=false,NPtsPath=100,NLayers=20)

    if PlotOn
        pygui(true)
        gcf()
    end
    PointPath_SingleLoop =  MakeEllip(r₁, r₂;NPts=NPtsPath)
    
    LMat = zeros(N, N)
    if PlotOn
        scatter3D(PointPath_SingleLoop[:,1], PointPath_SingleLoop[:,2], PointPath_SingleLoop[:,3])
    end
    NewPP = MakeEllip(r₁,r₂;Center = [0, 0, 1/ N * L],NPts = NPtsPath)

    if PlotOn
        scatter3D(NewPP[:,1], NewPP[:,2], NewPP[:,3])
    end

    Mut, Self, SavedBSelfArr = Mutual_L_TwoLoops(PointPath_SingleLoop, NewPP;DownSampleFac=DownSampleFac,MeasLayers=NLayers,MinThreshold=1e-10,IncludeWireInduct=false,SaveΦ₁=true)
    LMat .+= Self .* OffDiagOnes(N, 1)
    LMat .+= Mut .* OffDiagOnes(N, 2)
    for i in 3:N
        NewPP = MakeEllip(r₁,r₂;Center = [0, 0, (i-1)/ N * L],NPts = NPtsPath)

        if PlotOn
            scatter3D(NewPP[:,1], NewPP[:,2], NewPP[:,3])
        end
        Mut = Mutual_L_TwoLoops(PointPath_SingleLoop, NewPP, SavedBSelfArr;DownSampleFac=DownSampleFac,MeasLayers=NLayers,MinThreshold=1e-10)
        println(Mut)
        
        LMat .+= Mut .* OffDiagOnes(N, i)
        
        println(i)
    end
    LMat[N,N] = LMat[1,1]
    return LMat, sum(LMat)


end

"""
This function makes an elliptic solenoid and calculates the field at the center of the solenoid
        Lₘ₂₁ = Lₘ₁₂
        Total L = ∑ᵢ∑ⱼ(Lᵢⱼ)
    The inputs are:
    N - Number of turns
    r₁ - one radius of the ellipse
    r₂ - Second radius of ellipse
    L - length of solenoid (axially)
    kwargs:

    NPts_Coil  - number of points that each wire loop will be discretized into,
"""
function EvalField_Centered(
    N, #Number of turns
    r₁, #Radius 1 of ellipse
    r₂,#Radius 2 of ellipse
    L; #Length of solenoid in meters
    NPts_Coil=100
    )



            FieldCentered = [
                BiotSav(
                    MakeEllip(
                                r₁, #Radius 1 of ellipse
                                r₂;#Radius 2 of ellipse
                                Center = [0, 0, (Wᵢ - 1) / N * L], #The solenoid's axis is in Z
                                NPts = NPts_Coil, #How discretized the windings are
                                ),
                        [0, 0, L/2];
                        MinThreshold = 0.001,
                        )[3]
                        for Wᵢ in 1:N
                        ]

    return sum(FieldCentered)
end

"""
This function takes in a vector for each of the following:
    Number of turns (NVec) in each of the coupled solenoids
    r₁/₂ are the first and second radii of the elliptic cross section
    L is the length of each solenoid section
    Zoffset is the center offset distance of each solenoid. 

    The solenoids are built symmetrically so the Z offset is the center location
    
"""
function makeConcentricSolenoids(NVec,r₁Vec,r₂Vec,LVec,ZOffsetVec;
    DownSampleFac=1,PlotOn=false,NPtsPath=50,NLayers=15)

    NCoils = length(NVec)
    PP_List = vcat([[MakeEllipSolenoidPointPath(NVec[i],r₁Vec[i],r₂Vec[i],LVec[i],ZOffsetVec[i];NPtsPath=NPtsPath)] for i in 1:NCoils]...)
    LMat = zeros(NCoils,NCoils)

    if PlotOn
        for j in 1:NCoils
            pygui(true)
            plot3D(PP_List[j][:,1],PP_List[j][:,2],PP_List[j][:,3])
        end
    end

    for j in 1:NCoils
        OtherCoilsIndex = filter(x-> x!==j,1:NCoils) #Make a vector of ind of other coils

        for k in OtherCoilsIndex
            
            Mut, Self, SavedBSelfArr = Mutual_L_TwoLoops(PP_List[j], PP_List[k];DownSampleFac=DownSampleFac,MeasLayers=NLayers,MinThreshold=1e-10,IncludeWireInduct=false,SaveΦ₁=false)
            LMat[j,j] = Self
            LMat[j,k] = Mut
            LMat[k,j] = Mut
        end

    end
    

    return LMat,PP_List
end


function MakeEllipSolenoidPointPath(N,r₁,r₂,FullLength,Zoffset;
    NPtsPath=100)
    dL = FullLength / (N-1)  
    DriveCoilPointPath = vcat([MakeEllip(r₁, r₂; Center = [0, 0, L+Zoffset], NPts = NPtsPath) for L in (-FullLength/2:dL:FullLength/2)]...)  
    return DriveCoilPointPath
end

File: Biot.jl-main/src_python/MakeSolenoid.py
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D  # 3Dプロット用

from BiotSav import biot_sav
from MutualindductanceLoops import mutual_l_two_loops, vec_dist
from FieldCalc import make_ellip

def off_diag_ones(N, k):
    """
    OffDiagOnes(N, k): 
    k=1 → identity matrix; 
    k>=2 → ones on ±(k-1) off-diagonals.
    """
    M = np.zeros((N, N), dtype=float)
    offset = k - 1
    for i in range(N - offset):
        M[i, i + offset] = 1.0
        if offset != 0:
            M[i + offset, i] = 1.0
    return M

def eval_solenoid_induct(N, r1, r2, L, *,
                         down_sample_fac=1,
                         plot_on=False,
                         n_pts_path=100,
                         n_layers=20):
    """
    eval_Solenoid_Induct(N, r1, r2, L;
                         DownSampleFac=1, PlotOn=False,
                         NPtsPath=100, NLayers=20)
    → (L_mat, total_L)
    """
    # １ターン目ループ点列
    pp0 = make_ellip(r1, r2, n_pts=n_pts_path)

    # 全体行列の初期化
    L_mat = np.zeros((N, N), dtype=float)

    if plot_on:
        fig = plt.figure()
        ax = fig.add_subplot(111, projection='3d')
        ax.scatter(pp0[:,0], pp0[:,1], pp0[:,2], s=2)
        ax.set_title("Turn 1")

    # ２ターン目
    z2 = (1 / N) * L
    pp1 = make_ellip(r1, r2, center=[0, 0, z2], n_pts=n_pts_path)
    if plot_on:
        ax.scatter(pp1[:,0], pp1[:,1], pp1[:,2], s=2)

    # Self & first mutual
    mut, self_L, saved_phi1 = mutual_l_two_loops(
        pp0, pp1,
        down_sample_fac=down_sample_fac,
        meas_layers=n_layers,
        min_threshold=1e-10,
        include_wire_induct=False,
        save_phi1=True
    )
    L_mat += self_L * off_diag_ones(N, 1)
    L_mat += mut    * off_diag_ones(N, 2)

    # 残りのターン（3..N）
    for i in range(3, N+1):
        zi = ((i - 1) / N) * L
        ppi = make_ellip(r1, r2, center=[0, 0, zi], n_pts=n_pts_path)
        if plot_on:
            ax.scatter(ppi[:,0], ppi[:,1], ppi[:,2], s=2)

        # saved_phi1 を使った mutual
        mut = mutual_l_two_loops(
            pp0, ppi, saved_phi1,
            down_sample_fac=down_sample_fac,
            meas_layers=n_layers,
            min_threshold=1e-10
        )
        L_mat += mut * off_diag_ones(N, i)

    # 最後の対角要素を最初と揃える
    L_mat[N-1, N-1] = L_mat[0, 0]

    total_L = np.sum(L_mat)
    return L_mat, total_L

def eval_field_centered(N, r1, r2, L, *, n_pts_coil=100):
    """
    EvalField_Centered(N, r1, r2, L; NPts_Coil=100) → sum of B_z at center
    """
    vals = []
    for wi in range(1, N+1):
        zi = (wi - 1) / N * L
        pp = make_ellip(r1, r2, center=[0, 0, zi], n_pts=n_pts_coil)
        B = biot_sav(pp.tolist(), [0, 0, L/2], min_threshold=0.001)
        vals.append(B[2])
    return sum(vals)

def make_ellip_solenoid_point_path(N, r1, r2, full_length, zoffset, *,
                                   n_pts_path=100):
    """
    MakeEllipSolenoidPointPath(N, r1, r2, FullLength, Zoffset;
                               NPtsPath=100) → (N·NPtsPath×3) ndarray
    """
    dL = full_length / (N - 1)
    zs = np.linspace(-full_length/2, full_length/2, N)
    loops = [
        make_ellip(r1, r2, center=[0, 0, z + zoffset], n_pts=n_pts_path)
        for z in zs
    ]
    return np.vstack(loops)

def make_concentric_solenoids(N_vec, r1_vec, r2_vec, L_vec, zoffset_vec, *,
                              down_sample_fac=1,
                              plot_on=False,
                              n_pts_path=50,
                              n_layers=15):
    """
    makeConcentricSolenoids(NVec, r1Vec, r2Vec, LVec, ZOffsetVec;
                            DownSampleFac=1, PlotOn=False,
                            NPtsPath=50, NLayers=15)
    → (L_mat, pp_list)
    """
    N_coils = len(N_vec)
    pp_list = [
        make_ellip_solenoid_point_path(
            N_vec[i],
            r1_vec[i],
            r2_vec[i],
            L_vec[i],
            zoffset_vec[i],
            n_pts_path=n_pts_path
        )
        for i in range(N_coils)
    ]
    L_mat = np.zeros((N_coils, N_coils), dtype=float)

    if plot_on:
        fig = plt.figure()
        for j, pp in enumerate(pp_list, start=1):
            ax = fig.add_subplot(1, N_coils, j, projection='3d')
            ax.plot(pp[:,0], pp[:,1], pp[:,2], lw=1)

    for j in range(N_coils):
        for k in range(N_coils):
            if j == k:
                continue
            mut, self_L, _ = mutual_l_two_loops(
                pp_list[j], pp_list[k],
                down_sample_fac=down_sample_fac,
                meas_layers=n_layers,
                min_threshold=1e-10,
                include_wire_induct=False,
                save_phi1=False
            )
            L_mat[j, j] = self_L
            L_mat[j, k] = mut
            L_mat[k, j] = mut  # 対称

    return L_mat, pp_list
